File: LeetCode/Array/Majority_Element.cpp
/*Given an array of size n, find the majority element. The majority element is the element that appears more than ⌊ n/2 ⌋ times.

You may assume that the array is non-empty and the majority element always exist in the array.
*/

#include <iostream>
#include <vector>
using namespace std;

class Solution{
public:
    //Using Moore’s Voting Algorithm
    int majorityElement(vector<int> &num){
        int majority_index=0;
        int count =1;
        vector<int>::size_type it;
        for(it=0;it!=num.size();++it){
            if(num[majority_index]==num[it]){
                count++;
            }else{
                count--;
            }
            if(count==0){
                majority_index = it;
                count = 1;
            }
        }
        return num[majority_index];
    }
};

int main(){
    Solution solution;
    vector<int> v;
    v.push_back(1);
    v.push_back(2);
    v.push_back(1);
    v.push_back(1);
    v.push_back(3);
    v.push_back(2);
    v.push_back(1);
    cout<<solution.majorityElement(v)<<endl; //1
    return 0;
}

File: LeetCode_Python/Array/Majority_Element.py
# -*- coding: utf-8 -*

'''
Given an array of size n, find the majority element. 

The majority element is the element that appears more than ⌊ n/2 ⌋ times.

You may assume that the array is non-empty and the majority element always exist in the array.

'''

import math
class Solution(object):
	def majorityElement(self, nums):
		hash = {}
		for n in nums:
			if n not in hash.keys():
				hash[n] = 1
			else:
				temp = hash[n]
				hash[n] = temp + 1
		maximum = 0
		majority = nums[0]
		for i in hash:
			if hash[i] > maximum:
				maximum = hash[i]
				majority = i

		assert maximum > math.floor(len(nums)/2.0)
		return majority 

	# Moore投票算法
	def majorityElement2(self, nums):
		candicate, count = None, 0
		for n in nums:
			if count == 0:
				candicate, count = n, 1
			elif n == candicate:
				count += 1
			else:
				count -= 1
		return candicate

if __name__ == '__main__':
	nums = [1,2,1,1,3]
	result = Solution().majorityElement(nums)
	print result # 1